### PR TITLE
[REVIEW] Profile graph creation runtime and memory footprint 

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR closes  part of  https://github.com/rapidsai/graph_dl/issues/187 and https://github.com/rapidsai/cugraph/issues/3439 
Command to run:
```python3
python cugraph_graph_creation.py
```

See below benchmarks on SG (Works on MNMG too) : 

|scale|num_input_edges|directed|renumber|input_memory_per_worker|peak_allocation_workers|input_to_peak_ratio|output_to_peak_ratio|
|-----|---------------|--------|--------|-----------------------|------------------------------|-------------------|--------------------|
|22   |67108864       |TRUE    |TRUE    |1.0GB                  |5.9GB                         |5.885       |7.69         |
|22   |67108864       |TRUE    |FALSE   |1.0GB                  |5.9GB                         |5.885        |7.69         |
|22   |67108864       |FALSE   |TRUE    |1.0GB                  |12.1GB                        |12.08        |6.13         |
|22   |67108864       |FALSE   |FALSE   |1.0GB                  |12.1GB                        |12.08        |6.13         |
|23   |134217728      |TRUE    |TRUE    |2.0GB                  |11.8GB                        |5.914        |7.70         |
|23   |134217728      |TRUE    |FALSE   |2.0GB                  |11.8GB                        |5.914        |7.70         |
|23   |134217728      |FALSE   |TRUE    |2.0GB                  |24.3GB                        |12.16        |6.14         |
|23   |134217728      |FALSE   |FALSE   |2.0GB                  |24.3GB                        |12.16        |6.14         |
|24   |268435456      |TRUE    |TRUE    |4.0GB                  |23.8GB                        |5.93        |7.72         |
|24   |268435456      |TRUE    |FALSE   |4.0GB                  |23.8GB                        |5.93        |7.72         |

image

### CODE REVIEW

File: benchmarks/cugraph/standalone/cugraph_graph_creation.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    )
+    return (
+        input_to_peak_ratio,
+        output_to_peak_ratio,
+        input_memory_per_worker,
+        peak_allocation_across_workers,
+    )
+
+
+def restart_client(client):
```
Comment (rlratzel) [inline benchmarks/cugraph/standalone/cugraph_graph_creation.py:113]: I don't have a strong opinion, but I wonder if this should go in `testing.mg_utils` too.

File: python/cugraph/cugraph/testing/mg_utils.py [hunk 2]
```diff
@@ -77,3 +86,185 @@ def stop_dask_client(client, cluster=None):
...
+        scale,
+        (2**scale) * edgefactor,
+        0.57,  # from Graph500
+        0.19,  # from Graph500
+        0.19,  # from Graph500
+        seed or 42,
+        clip_and_flip=False,
+        scramble_vertex_ids=True,
+        create_using=None,  # return edgelist instead of Graph instance
+        mg=True,
```
Comment (rlratzel) [inline python/cugraph/cugraph/testing/mg_utils.py:131]: Should this be a parameter to the function to allow for both SG `cudf` and MG `dask_cudf` return types too?

File: benchmarks/cugraph/standalone/cugraph_graph_creation.py (hunk 1 above)
Comment (VibhuJawa) [reply benchmarks/cugraph/standalone/cugraph_graph_creation.py:113]: Yup, I should move it there.

File: python/cugraph/cugraph/testing/mg_utils.py [hunk 3]
```diff
@@ -77,3 +86,185 @@ def stop_dask_client(client, cluster=None):
...
     print("\ndask_client closed.")
+
+
+def enable_spilling():
+    import cudf
+
+    cudf.set_option("spill", True)
+
+
+def generate_edgelist(
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py:97]: I will change the name to `generate_edgelist_rmat`

File: python/cugraph/cugraph/testing/mg_utils.py [hunk 4]
```diff
@@ -77,3 +86,185 @@ def stop_dask_client(client, cluster=None):
...
+                return return_val, allocation_counts
+            else:
+                return return_val
+
+        return wrapper
+
+    return decorator
+
+
+def _print_allocation_statistics(func, args, kwargs, execution_time, allocation_counts):
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py:253]: Will move it into a function and then call print

File: python/cugraph/cugraph/testing/mg_utils.py [hunk 5]
```diff
@@ -77,3 +86,185 @@ def stop_dask_client(client, cluster=None):
...
+    print(f"function:  {func.__name__}\n")
+    print(f"function args: {args} kwargs: {kwargs}", flush=True)
+    print(f"execution_time: {execution_time}", flush=True)
+    print("allocation_counts")
+    allocation_counts_parsed = {
+        worker_id: _parse_allocation_counts(worker_allocations)
+        for worker_id, worker_allocations in allocation_counts.items()
+    }
+    pprint(allocation_counts_parsed, indent=4, width=1, compact=True)
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py]: Resolved by https://github.com/rapidsai/cugraph/pull/3518/commits/005b698413ad4f3b849461831175ec75f04533f1

File: python/cugraph/cugraph/testing/mg_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py]: Resolved by https://github.com/rapidsai/cugraph/commit/005b698413ad4f3b849461831175ec75f04533f1

File: python/cugraph/cugraph/testing/mg_utils.py (hunk 2 above)
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py]: Resolved by https://github.com/rapidsai/cugraph/commit/005b698413ad4f3b849461831175ec75f04533f1

File: python/cugraph/cugraph/testing/mg_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/testing/mg_utils.py]: Resolved by https://github.com/rapidsai/cugraph/commit/005b698413ad4f3b849461831175ec75f04533f1

File: benchmarks/cugraph/standalone/cugraph_graph_creation.py (hunk 1 above)
Comment (VibhuJawa) [reply benchmarks/cugraph/standalone/cugraph_graph_creation.py]: Resolved by https://github.com/rapidsai/cugraph/commit/005b698413ad4f3b849461831175ec75f04533f1

**Review by rlratzel — APPROVED** [review]
Nice!

**Pull request merged**